### PR TITLE
Feature island optimizer

### PR DESCRIPTION
**Pull request by cliff-bohm**
new (meta) optimizer which creates islands of organisms each with its own simple optimizer and allows for migration between islands.

### CODE REVIEW

**Comment by cgnitash** [comment]
This is very cool. I assume this allows me to _always_ use island-optimizer. So every experiment would use island-optimizer as its top-level optimizer, and If I only need a single optimizer then the default is islandnameSpaceList = ["root::"].
Is that the broad idea?

File: Optimizer/IslandsOptimizer/IslandsOptimizer.h [hunk 1]
```diff
@@ -0,0 +1,210 @@
...
+
+
+class IslandsOptimizer : public AbstractOptimizer {
+public:
+
+	static std::shared_ptr<ParameterLink<std::string>> IslandNameSpaceListPL;
+	static std::shared_ptr<ParameterLink<double>> migrationRatePL;
+
+	std::vector <std::shared_ptr<AbstractOptimizer>> islandOptimizers;
+	int islands;
```
Comment (cgnitash) [inline Optimizer/IslandsOptimizer/IslandsOptimizer.h]: Just make this size_t. Thenyou can avoid the static_cast in the .cpp

File: Optimizer/IslandsOptimizer/IslandsOptimizer.cpp [hunk 2]
```diff
@@ -0,0 +1,85 @@
...
+	"OPTIMIZER_ISLANDS-migrationRate", .02,
+	"% of new organisms which migrate to a random island at birth");
+
+IslandsOptimizer::IslandsOptimizer(std::shared_ptr<ParametersTable> PT_)
+    : AbstractOptimizer(PT_) {
+
+	std::vector<std::string> opNameSpaces;
+	convertCSVListToVector(IslandNameSpaceListPL->get(PT), opNameSpaces);
+
+	for (auto nameSpace : opNameSpaces) {
```
Comment (cgnitash) [inline Optimizer/IslandsOptimizer/IslandsOptimizer.cpp]: This should be auto&. There's no need to copy the strings. Ditto on line 36.

**Comment by JorySchossau** [comment]
I think making Island Optimization the only optimizer is not a good idea, because while one could think of a typical MABE experiment as a single island it would be conceptually and programmatically annoying. Currently, I think we figure using islands is not as common as single population runs, so single population runs are easier / default.

File: Optimizer/IslandsOptimizer/IslandsOptimizer.h (hunk 1 above)
Comment (cliff-bohm) [reply Optimizer/IslandsOptimizer/IslandsOptimizer.h]: I changed this to size_t, but I still needed one static cast when writing to data map.

**Comment by cliff-bohm** [comment]
changes made

**Comment by cgnitash** [comment]
Ok, that seems reasonable. We can keep that idea in the back of our minds for redesign. Maintaining ease of use for common cases is a good priority for the moment.

**Comment by cgnitash** [comment]
Works fine. That is, it compiles, and doesn't crash for the bunch of parameter combos I tested.

**Pull request merged**